proto/backend.proto

```diff
@@ -0,0 +1,212 @@
+syntax = "proto3";
+import "google/protobuf/empty.proto";
+package grafana.plugin;
+
+//-----------------------------------------------
+// Common
+//-----------------------------------------------
+
+message PluginConfig {
+  int64 id = 1;
+  int64 orgId = 2;
+  string name = 3;
+  string type = 4;
+  string url = 5;
+  string jsonData = 6;
+  map<string,string> decryptedSecureJsonData = 7;
+}
+
+
+//-----------------------------------------------
+// All Backend Plugins Implement
+//-----------------------------------------------
+
+service BackendPlugin {
+  // Get the plugin status
+  rpc Check(PluginStatusRequest) returns (PluginStatusResponse);
+
+  // HTTP Style request
+  rpc Resource(ResourceRequest) returns (ResourceResponse);
+
+  // Well typed query interface
+  rpc DataQuery(DataQueryRequest) returns (DataQueryResponse);
+
+}
+
+//-----------------------------------------------
+// Plugin Status Check
+//-----------------------------------------------
+
+message PluginStatusRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  enum FetchInfo {
+    STATUS = 0;  // nothing
+    API = 1;     // OpenAPI description
+    METRICS = 2; // Prometheus style metrics
+    DEBUG = 3;   // JSON Debug Info (admin+ view)
+  }
+
+  // Get an OpenAPI description (viewer+)
+  FetchInfo fetch = 3;
+}
+
+message PluginStatusResponse {
+  enum PluginStatus {
+    UNKNOWN = 0; // starting etc
+    OK = 1;
+    ERROR = 2;
+  }
+
+  PluginStatus status = 1;
+  string info = 2;
+}
+
+//-----------------------------------------------
+// Data Query
+//-----------------------------------------------
+
+message TimeRange {
+  int64 fromEpochMS = 1;
+  int64 toEpochMS = 2;
+}
+
+message DataQuery {
+  string refId = 1;
+  int64 maxDataPoints = 2;
+  int64 intervalMS = 3;
+  TimeRange timeRange = 4; // NOTE: this moved from QueryRequest in v1
+  bytes json = 5; // Arbitrary structure (model?  modelJson in v1)
+}
+
+message DataQueryRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  // List of queries
+  repeated DataQuery queries = 3;
+}
+
+message DataQueryResponse {
+  // Arrow encoded DataFrames
+  // Each frame encodes its own: Errors, meta, and refId
+  repeated bytes frames = 1;
+
+  // Additional response metadata
+  map<string,string> metadata = 2;
+}
+
+//-----------------------------------------------
+// HTTP style request
+//-----------------------------------------------
+
+message ResourceRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  
+  // HTTP Style parameters
+  map<string,string> headers = 2; // Request Headers
+  string method = 3; // GET/POST/DELETE/etc
+  string path = 4;   // (with params?)
+  bytes body = 5;    // assume string unless header says otherwise?
+}
+
+message ResourceResponse {
+  int32 code = 1;  // Match HTTP status codes
```

Comment:
*Must* this match HTTP status codes? If that is the case, will make a more specific type in the SDK. 

If it is meant to be more open ended, then maybe a string is better.

Reply:
I think it should be passed back directly as the HTTP status code -- and appropriate behavior should be triggered off the standard codes.  If we want to enumerate the codes or have it open ended, I don't have a strong opinion.